### PR TITLE
fix: Curation API - fix `PATCH /collections/{collection_id}`

### PR DESCRIPTION
**Pull request by danieljhegeman**
This work was largely already approved in [3028](https://github.com/chanzuckerberg/single-cell-data-portal/pull/3028), but I reverted that due to smoke test failures (turned out I wasn't responsible). 

Anyways, this needs to be added back in, with the _additional_ change that centers around making sure that dois are handled appropriately:

@brianraymor 👀 :
- do not delete doi or metadata (no change) when a bad doi is submitted (this differs from the UI!); just return 400 
- do not delete doi if no new doi is submitted
- return `publisher_metadata` in 200 response

- #3010 

### Reviewers
**Functional:** 
@Bento007 
**Readability:** 
@nayib-jose-gloria 
---

## Changes
- add
- remove
- modify

## QA steps (optional)
- presently testing on dev

### CODE REVIEW

**Comment by brianraymor** [comment]
> do not delete doi or metadata (no change) when a bad doi is submitted (this differs from the UI!); just return 400

What constitutes a _bad doi_? And how is this messaged to the curator?

**Comment by danieljhegeman** [comment]
@brianraymor @jahilton do we want to return `publisher_metadata` in the response, since this is impacted when the doi is updated?

**Comment by danieljhegeman** [comment]
> > do not delete doi or metadata (no change) when a bad doi is submitted (this differs from the UI!); just return 400
> 
> What constitutes a _bad doi_? And how is this messaged to the curator?
> 
> Might also be good to get @ebezzi 's eyes on this since he implemented the current DOI changes in the UX.

@brianraymor a doi that cannot be found in Crossref is a bad doi in this context. Generally speaking there are two categories, each of which has a slightly different 400 error that is returned: 
- malformed - returns `"Invalid DOI"`
- cannot be found in Crossref (yet) - returns `"DOI cannot be found on Crossref"`

**Comment by brianraymor** [comment]
What is the format for the DOI?

**Comment by danieljhegeman** [comment]
> What is the format for the DOI? For example, is it a DOI curie?
> 
> I'm curious because in the UX we now only allow curators to type in the unique identifier prompted by `doi:`.

It accepts `doi:<doi>` or `http[s]://doi.org/<doi>`

So `doi:10.1523/JNEUROSCI.1720-15.2016`, `http://doi.org/10.1523/JNEUROSCI.1720-15.2016`, or `https://doi.org/10.1523/JNEUROSCI.1720-15.2016`.

File: backend/corpora/lambdas/api/v1/curation/collections/collection_id/actions.py [hunk 1]
```diff
@@ -35,5 +50,57 @@ def get(collection_id: str, token_info: dict):
...
+                    if link.link_type == CollectionLinkType.DOI
+                ]
+            )
+    if errors:
+        raise InvalidParametersHTTPException(detail=errors)
+
+    collection.update(**body, keep_links=keep_links)
+    collection_dict = collection.to_dict()
+    collection_dict["links"] = [
+        dict(link_url=link["link_url"], link_name=link.get("link_name", ""), link_type=link["link_type"])
```
Comment (Bento007) [inline backend/corpora/lambdas/api/v1/curation/collections/collection_id/actions.py:103]: Why is this needed?

File: backend/corpora/lambdas/api/v1/curation/collections/collection_id/actions.py [hunk 2]
```diff
@@ -35,5 +50,57 @@ def get(collection_id: str, token_info: dict):
...
+                    link.to_dict_keep({DbCollectionLink: EntityColumns.link_cols})
+                    for link in collection.links
+                    if link.link_type == CollectionLinkType.DOI
+                ]
+            )
+    if errors:
+        raise InvalidParametersHTTPException(detail=errors)
+
+    collection.update(**body, keep_links=keep_links)
+    collection_dict = collection.to_dict()
```
Comment (Bento007) [inline backend/corpora/lambdas/api/v1/curation/collections/collection_id/actions.py:101]: if you use to_dict_keep you wont need to do as much data shuffling below. You'll also avoid converting the whole collection to dict.

File: tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py [hunk 3]
```diff
@@ -409,16 +410,118 @@ def test__update_collection__OK(self):
...
+
+    def test__update_collection__links_and_doi_management__OK(self):
+        name = "partial updates test collection"
+        links = [
+            {"link_name": "name", "link_type": "RAW_DATA", "link_url": "http://test_link.place"},
+            {"link_name": "doi", "link_type": "DOI", "link_url": "http://doi.doi"},
+        ]
+        new_links = [
+            {"link_name": "new link", "link_type": "RAW_DATA", "link_url": "http://brand_new_link.place"},
+            {"link_name": "new doi", "link_type": "DOI", "link_url": "http://doi.org/10.1016"},  # a real DOI
```
Comment (Bento007) [inline tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py:458]: mock this out rather than hit the DOI server. Otherwise, this could turn into another flakey API like dropbox..

File: tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py [hunk 4]
```diff
@@ -409,16 +410,118 @@ def test__update_collection__OK(self):
...
+
+        for test_title, initial_links, new_links, expected_status_code, expected_links in links_configurations:
+            with self.subTest(test_title):
+                collection_id = self.generate_collection(self.session, name=name, links=initial_links).id
+                original_collection = self.app.get(f"curation/v1/collections/{collection_id}").json
+                self.assertEqual(initial_links if initial_links else [], original_collection["links"])
+                metadata = {"links": new_links} if new_links is not None else {}
+                response = self.app.patch(
+                    f"/curation/v1/collections/{collection_id}",
+                    data=json.dumps(metadata),
```
Comment (Bento007) [inline tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py:480]:
```suggestion
                    json=metadata,
```

File: tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py [hunk 5]
```diff
@@ -409,16 +410,118 @@ def test__update_collection__OK(self):
...
+
+    def test__update_collection__doi_does_not_exist__BAD_REQUEST(self):
+        name = "bad doi update test collection"
+        links = [
+            {"link_name": "name", "link_type": "RAW_DATA", "link_url": "http://test_link.place"},
+            {"link_name": "doi", "link_type": "DOI", "link_url": "http://doi.doi"},
+        ]
+        new_links_bad_doi = [
+            {"link_name": "new link", "link_type": "RAW_DATA", "link_url": "http://brand_new_link.place"},
+            {"link_name": "new doi", "link_type": "DOI", "link_url": "http://a_bad_doi"},  # a bad DOI
```
Comment (Bento007) [inline tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py:496]: Make sure the request to the DOI sever is mocked out.

**Review by Bento007 — CHANGES_REQUESTED** [review]
suggestion: We should add the endpoint `put` collection which does do a full replacement. That way users could remove links if they desire.

File: backend/config/curation-api.yml [hunk 6]
```diff
@@ -164,6 +165,7 @@ paths:
         content:
           application/json:
             schema:
+              additionalProperties: false
```
[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/config/curation-api.yml:168]: No, it is not. I remove it in this PR so that I can use this portion of schema on the response for this endpoint. Otherwise, `additionalProperties: false` conflicts with the addition of `publisher_metadata`, which is now being returned.

File: backend/corpora/lambdas/api/v1/curation/collections/collection_id/actions.py [hunk 7]
```diff
@@ -35,5 +50,57 @@ def get(collection_id: str, token_info: dict):
     return jsonify(collection_response)
 
 
-def patch(collection_id: str, body: dict, token_info: dict):
-    return update_collection(collection_id, body, token_info)
+@dbconnect
+def patch(collection_id: str, body: dict, token_info: dict) -> Response:
+    """
+    Curation API: update the metadata for a Collection. Only included field will be altered. If links are provided, all
```
[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/corpora/lambdas/api/v1/curation/collections/collection_id/actions.py:57]: I think what you mean is... it _should_ be documented there 😂 😭 

Fixing...

File: backend/corpora/lambdas/api/v1/curation/collections/collection_id/actions.py [hunk 8]
```diff
@@ -35,5 +50,57 @@ def get(collection_id: str, token_info: dict):
...
+        for link in body["links"]:
+            if link["link_type"] == ProjectLinkType.DOI.name:
+                new_doi_provided = True
+    else:
+        keep_links = True  # links have NOT been provided; keep existing links
+    collection, errors = get_collection_and_verify_body(db_session, collection_id, body, token_info)
+
+    if new_doi_provided:
+        # Compute the diff between old and new DOI
+        old_doi = collection.get_doi()
```
[the remark this answers is not in the document]
Comment (danieljhegeman) [reply backend/corpora/lambdas/api/v1/curation/collections/collection_id/actions.py:72]: No, there are intervening lines of logic in the code you linked to that we do not want here. It is part of why I am not reusing that other update handler function wholly.

File: backend/corpora/lambdas/api/v1/curation/collections/collection_id/actions.py (hunk 1 above)
Comment (danieljhegeman) [reply backend/corpora/lambdas/api/v1/curation/collections/collection_id/actions.py:94]: I think I should have removed this.

File: tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py (hunk 5 above)
Comment (danieljhegeman) [reply tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py:492]: It doesn't need to be. It is handled by the existing code. The request results in a return value of `None` for the new doi; this is handled correctly by the existing algorithm in the patch request handler.

Comment (danieljhegeman) [reply tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py:492]: It is handled below by the `with patch(...)` code on ln 500

File: tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py (hunk 4 above)
Comment (danieljhegeman) [reply tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py:476]: that doesn't work. It requires JSON.

**Review by Bento007 — APPROVED** [review]

File: tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py (hunk 4 above)
Comment (Bento007) [reply tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py:476]: it does work https://github.com/chanzuckerberg/single-cell-data-portal/blob/d5423ef0bdb75f8b345a302d16875f2a42f5dfbc/tests/unit/backend/corpora/api_server/curator/collection/test_dataset.py#L69

Comment (danieljhegeman) [reply tests/unit/backend/corpora/api_server/curator/collection/test_curation_collections.py:476]: Ah, sorry, I tried it with `data=`.
